r1cs/gadgets/crypto/src/nizk/groth16/mod.rs

```diff
@@ -220,6 +293,96 @@ for VerifyingKeyGadget<PairingE, ConstraintF, P>
                     P::G1Gadget::alloc(cs.ns(|| format!("gamma_abc_{}", i)), || {
                         Ok(gamma_abc_i.into_projective())
                     })
+                        .and_then(|input_g| {
+                            input_g
+                                .is_zero(cs.ns(|| format!("is input {} zero", i)))?
+                                .enforce_equal(
+                                    cs.ns(|| format!("input {} must not be zero", i)),
+                                    &Boolean::constant(false),
+                                )?;
+                            Ok(input_g)
+                        })
+                })
+                .collect::<Vec<_>>()
+                .into_iter()
+                .collect::<Result<_, _>>()?;
+            Ok(Self {
+                alpha_g1_beta_g2,
+                gamma_g2,
+                delta_g2,
+                gamma_abc_g1,
+            })
+        })
+    }
+
+    #[inline]
+    fn alloc_checked<FN, T, CS: ConstraintSystem<ConstraintF>>(
```

Comment:
As for `alloc()`: I cant think of any application where a verification key is a private circuit witness.